sources/directconnect/direct-connect-gateway-attachment.go

```diff
@@ -0,0 +1,119 @@
+package directconnect
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/directconnect"
+	"github.com/overmindtech/aws-source/sources"
+	"github.com/overmindtech/sdp-go"
+)
+
+func directConnectGatewayAttachmentOutputMapper(_ context.Context, _ *directconnect.Client, scope string, _ *directconnect.DescribeDirectConnectGatewayAttachmentsInput, output *directconnect.DescribeDirectConnectGatewayAttachmentsOutput) ([]*sdp.Item, error) {
+	items := make([]*sdp.Item, 0)
+
+	for _, attachment := range output.DirectConnectGatewayAttachments {
+		attributes, err := sources.ToAttributesCase(attachment, "tags")
+		if err != nil {
+			return nil, err
+		}
+
+		// The uniqueAttributeValue for this is a custom field:
+		// {gatewayId} {virtualInterfaceId}
+		// i.e., "cf68415c-f4ae-48f2-87a7-3b52cexample dxvif-ffhhk74f"
+		err = attributes.Set("uniqueName", fmt.Sprintf(gatewayIDVirtualInterfaceIDFmt, *attachment.DirectConnectGatewayId, *attachment.VirtualInterfaceId))
+		if err != nil {
+			return nil, err
+		}
+
+		item := sdp.Item{
+			Type:            "directconnect-direct-connect-gateway-attachment",
+			UniqueAttribute: "uniqueName",
+			Attributes:      attributes,
+			Scope:           scope,
+		}
+
+		if attachment.DirectConnectGatewayId != nil {
+			// +overmind:link directconnect-direct-connect-gateway
+			item.LinkedItemQueries = append(item.LinkedItemQueries, &sdp.LinkedItemQuery{
+				Query: &sdp.Query{
+					Type:   "directconnect-direct-connect-gateway",
+					Method: sdp.QueryMethod_GET,
+					Query:  *attachment.DirectConnectGatewayId,
+					Scope:  "global",
+				},
+				BlastPropagation: &sdp.BlastPropagation{
+					// This is not clearly documented, but it seems that if the gateway is deleted, the attachment state will change to detaching
+					In: true,
+					// We can't affect the direct connect gateway
+					Out: false,
+				},
+			})
+		}
+
+		if attachment.VirtualInterfaceId != nil {
+			// +overmind:link directconnect-virtual-interface
+			item.LinkedItemQueries = append(item.LinkedItemQueries, &sdp.LinkedItemQuery{
+				Query: &sdp.Query{
+					Type:   "directconnect-virtual-interface",
+					Method: sdp.QueryMethod_GET,
+					Query:  *attachment.VirtualInterfaceId,
+					Scope:  scope,
+				},
+				BlastPropagation: &sdp.BlastPropagation{
+					// If virtual interface is deleted, the attachment state will change to detaching
+					// https://docs.aws.amazon.com/directconnect/latest/APIReference/API_DirectConnectGatewayAttachment.html#API_DirectConnectGatewayAttachment_Contents
+					In: true,
+					// We can't affect the virtual interface
+					Out: false,
```

Comment:
I think in this instance we would want the blast to propagate in both directions. If you imagine a graph which is `direct-connect-gateway` -> `direct-connect-gateway-attachment` -> `directconnect-virtual-interface`

If we were to change the gateway, the blast radius wouldn't include the interface, because the outbound blast prop is false. I think realistically if you were to change a direct connect connections it could affect the interfaces and the things that use them couldn't it?

```suggestion
					Out: true,
```

Reply:
Virtual interface <-> gateway relation is defined on the virtual interface source: https://github.com/tomasdembelli/aws-source/blob/main/sources/directconnect/virturl-interface.go#L61

So, changing a gateway would affect the associated virtual interfaces.

My understanding is that the attachment source is just a state source. There is no create or delete method for this source. So, it can only change if the associated sources (virtual interface and gateway) are changed.
It doesn't have a terraform source and a List method, so defining the gateway <-> interface relation on the sources itself seemed more robust to me.

But I see your point, in the graph, maybe it is better so see the gateway to interface connection over their attachment.
In that case we need to remove the relation between virtual interface and gateway on the interface (and soon on the gateway) source.

I can do the change (and as you mentioned in the other relevant comments), I just want to make sure we are on the same page.

Reply:
Ah you're right, I hadn't noticed the direct relationship. In that case this implementation is good